src/wallet/silentpayments.h

```diff
@@ -0,0 +1,178 @@
+ *
+ * @param tx_outpoints      The transaction outpoints.
+ * @param sum_input_pubkeys The sum of the eligible input public keys.
+ * @return uint256          The computed hash value.
+ *
+ * @see ComputeInputsHash(const std::vector<COutPoint>& tx_outpoints, const CPubKey& sum_input_pubkeys);
+ */
+uint256 ComputeInputsHash(const std::vector<COutPoint>& tx_outpoints, const CPubKey& sum_input_pubkeys);
+
+/**
+ * @brief Compute the shared secret tweak.
+ *
+ * In a single transaction, a sender can create multiple outputs for the same recipient. To prevent
+ * needing to do an ECDH for each output, the ECDH public key is calculated once for the transaction
+ * and then hashed with the counter k. The outputs must be created sequentially starting with `k=0`.
+ *
+ * @param ecdh_pubkey The ECDH public key.
+ * @param k The output counter.
+ * @return uint256 The shared secret tweak for output_k.
+ */
+uint256 ComputeSharedSecretTweak(const CPubKey& ecdh_pubkey, const uint32_t k);
+
+/**
+ * @brief Compute the silent payment label tweak.
+ *
+ * The receiving wallet may hand out labeled silent payment addresses to determine the origin of a payment.
+ * Labels are specified by hashing the receivers scan key with a counter m. `m=0` is reserved for generating
+ * change outputs for the receiver. Additional labels can be specified starting from `m=1`. The counter m
+ * is hashed with the secret scan key to prevent an external observer from learning anything about the
+ * receiver's labels.
+ *
+ * It is strongly recommended labels be created sequentially starting from `m=1`. Doing so ensures that
+ * label data is not critical for backups, as the receiving wallet can pregenerate labels starting from 1
+ * up to some sufficiently large M when recovering from a backup.
+ *
+ * @param scan_key The secret scan key.
+ * @param m Integer representing the "m-th" label. m=0 is reserved for the change output.
+ * @return uint256 The computed silent payment label tweak.
+ */
+uint256 ComputeSilentPaymentLabelTweak(const CKey& scan_key, const int m);
+
+/**
+ * @brief Generate a silent payment labeled address.
+ *
+ * @param receiver The receiver's silent payment destination (i.e. scan and spend public keys).
+ * @param label The label tweak
+ * @return V0SilentPaymentDestination The slient payment destination, with `B_spend -> B_spend + label * G`.
+ *
+ * @see ComputeSilentPaymentLabelTweak(const CKey& scan_key, const int m);
+ */
+V0SilentPaymentDestination GenerateSilentPaymentLabeledAddress(const V0SilentPaymentDestination& receiver, const uint256& label);
+
+/**
+ * @brief Generate silent payment taproot destinations.
+ *
+ * Given a set of silent payment destinations, generate the requested number of outputs. If a silent payment
+ * destination is repeated, this indicates multiple outputs are requested for the same recipient. The silent payment
+ * desintaions are passed in map where the key indicates their desired position in the final tx.vout array.
```

Comment:
Not sure protocol implementation needs to now about the position in tx.vout array. AFAIK, output calculation doesn't depend on the position. Have you considered passing destinations one by one together with the counter if a destination is repeated?

Reply:
Output calculation doesn't depend on the position, but using output position makes the implementation much simpler. I did try something similar to what you suggested but it ended up being really complicated to make sure I was matching up all the right amounts and was always _implicitly_ relying on output position. @achow101 pointed out it would be better and simpler to just use the output index explicitly.

This also doesn't create any dependence on output position. The user can still order the outputs however they want after generating the silent payment scriptPubKeys.